ocaml/xapi/xapi_database_backup.ml

```diff
@@ -0,0 +1,175 @@
+    ) row.value
+
+let unpack_table __context (t:table) =
+  List.iter (unpack_row __context t t.tblname) (if t.value = [] then [empty_row] else t.value)
+
+let apply_updates __context updates =
+  List.iter (unpack_table __context) updates
+
+let get_table_list db =
+  Db_cache_types.TableSet.fold_over_recent (-2L) (fun table stat value acc -> table::acc) (Db_cache_types.Database.tableset db) []
+
+let counter db table =
+  let t = Db_cache_types.TableSet.find table (Db_cache_types.Database.tableset db) in
+  {tblname=table; count=0; del=(Db_cache_types.Table.get_deleted_len t)}
+
+let object_count db =
+  let tables = get_table_list db in
+  List.map (counter db) tables
+
+let count_eq_check c1 c2 =
+  c1.tblname = c2.tblname && c1.count = c2.count && c1.del = c2.del
+
+let apply_changes (delta:delta) __context =
+  if (Db_cache_types.Manifest.generation (Db_cache_types.Database.manifest !Xapi_slave_db.slave_db)) < delta.fresh_token then
+    begin
+      Xapi_slave_db.slave_db := Database.set_generation delta.fresh_token !Xapi_slave_db.slave_db;
+      apply_updates __context delta.tables;
+      apply_deletes __context delta.deletes;
+    end;
+  (* Need to check that we haven't missed anything *)
+  if not (List.for_all2 count_eq_check (object_count !Xapi_slave_db.slave_db) (object_count (Db_ref.get_database (Context.database_of __context)))) then
+    Mutex.execute Xapi_slave_db.slave_db_mutex (fun () -> Xapi_slave_db.slave_db := Db_cache_types.Database.set_generation 0L !Xapi_slave_db.slave_db)
+
+let _info_extract (tlist:table list) =
+  let tbls = List.map (fun (i:table) -> i.tblname) tlist in
+  let rows = List.concat (List.map (fun tbl -> List.map (fun row -> row.objref) tbl.value) tlist) in
+  List.concat (List.map ((fun objlist tblname -> List.combine [tblname] objlist) rows) tbls)
+
+let fl tblname rlist =
+  List.fold_left (fun acc r -> [{tbl=tblname; row=r.objref}] @ acc) [] rlist
+
+let info_extract (tlist:table list) =
+  List.fold_left (fun pl (t:table) -> (fl t.tblname t.value) @ pl) [] tlist
+
+let handle_notifications __context (d:delta) =
+  List.iter ((fun __context (d:del_tables) -> send_notification __context "del" {tbl=d.table; row=d.key}) __context) d.deletes;
+  List.iter (send_notification __context "mod") (info_extract d.tables)
+
+let endless_loop ~__context () =
+  let uri = Constants.database_backup_uri in
+  while (true) do
+    let addr = Pool_role.get_master_address() in
+    let psec = !Xapi_globs.pool_secret in
+    let token_str = Int64.to_string (Manifest.generation (Database.manifest !Xapi_slave_db.slave_db)) in
+    let req = Xapi_http.http_request ~cookie:["pool_secret", psec] ~query:[("token", token_str)] Http.Get uri in
+    let transport = SSL(SSL.make (), addr, !Xapi_globs.https_port) in
+    with_transport transport
+      (with_http req
+         (fun (response, fd) ->
+            let res = match response.Http.Response.content_length with
+              | None -> raise Content_length_required
```

Comment:
This exception, for example, is unhandled and prevents any further local DB updates until you restart xapi.